src/annotate/onclass/script.py

```diff
@@ -0,0 +1,209 @@
+    
+    Parameters
+    ----------
+    cl_obo_file : str
+        Path to the cell ontology file.
+        
+    Returns
+    -------
+    Tuple[Dict[str, str], Dict[str, str]]
+        A tuple of two dictionaries. The first dictionary maps cell ontology IDs to cell type names.
+        The second dictionary maps cell type names to cell ontology IDs.
+    """
+    graph = obonet.read_obo(cl_obo_file)
+    cl_id_to_name = {id_: data.get("name") for id_, data in graph.nodes(data=True)}
+    name_to_cl_id = {data.get("name"): id_ for id_, data in graph.nodes(data=True)}
+    return cl_id_to_name, name_to_cl_id
+
+def predict_input_data(model: OnClassModel,
+                       input_matrix: np.array,
+                       input_modality: ad.AnnData,
+                       id_to_name: dict,
+                       obs_prediction: str,
+                       obs_probability: str) -> ad.AnnData:
+    """
+    Predict cell types for input data and save results to Anndata obj.
+    
+    Parameters
+    ----------
+    model : OnClassModel
+        The OnClass model.
+    input_matrix : np.array
+        The input data matrix.
+    input_modality : ad.AnnData
+        The input data Anndata object.
+    id_to_name : dict
+        Dictionary mapping cell ontology IDs to cell type names.
+    obs_prediction : str
+        The obs key for the predicted cell type.
+    obs_probability : str
+        The obs key for the predicted cell type probability.
+        
+    Returns
+    -------
+    ad.AnnData
+        The input data Anndata object with the predicted cell types saved in obs.
+    """
+    corr_test_feature = model.ProcessTestFeature(
+        test_feature=input_matrix,
+        test_genes=input_modality.var_names,
+        log_transform=False,
+    )
+    onclass_pred = model.Predict(corr_test_feature, use_normalize=False, refine=True, unseen_ratio=-1.0)
+    pred_label = [model.i2co[ind] for ind in onclass_pred[2]]
+    pred_cell_type_label = [id_to_name[id] for id in pred_label]
+    
+    input_modality.obs[obs_prediction] = pred_cell_type_label
+    input_modality.obs[obs_probability] = np.max(onclass_pred[1], axis=1) / onclass_pred[1].sum(1)
+    return input_modality
+
+def main():
+    logger.info("Reading input data")
+    input_mudata = mu.read_h5mu(par["input"])
+    input_modality = input_mudata.mod[par["modality"]].copy()
+
+    id_to_name, name_to_id = map_celltype_to_ontology_id(par["cl_obo_file"])
+    obs_predictions = par["output_obs_predictions"] if par["output_obs_predictions"] else [f"{target}_pred" for target in par["reference_obs_targets"]]
+    obs_probabilities = par["output_obs_probability"] if par["output_obs_probability"] else [f"{target}_prob" for target in par["reference_obs_targets"]]
+    
+    if par["input_layer"]:
+        input_matrix = input_modality.layers[par["input_layer"]].toarray()
+    else:
+        input_matrix = input_modality.X.toarray()
+        
+    if par["reference"]:
```

Comment:
Can we turn the if/elif block around such that if a model is provided it will always be used?
```
if par["model]: 
    if par["reference]:
        log a warning that reference won't be used
    ...
elif par["reference"]:
    ...
```

Reply:
made it in the same way as celltypist - either model XOR reference has to be provided